README.md
# Ubuntu WSL Integration

This is Ubuntu WSL Integration (`ubuntu-wsl-integration`) for use with independently or with Ubuntu WSL Onboarding Experience (`ubuntu-wsl-oobe`).

This is proposed to ship with Ubuntu WSL metapackage in 20.10 under Ubuntu WSL meta-package (`ubuntu-wsl`).

This contains:

- Ubuntu WSL Integration script that can be used when startup or with `wslusc` utility in `wslu`;
- A cli for managing the integration.
- A Text-based UI for easier integration menagement.

## Bugs

Please report bugs to launchpad here: <https://bugs.launchpad.net/ubuntu-wsl-integration>

## License

GPLv3



debian/changelog
ubuntu-wsl-integration (0.31.3) impish; urgency=medium

  * wsl-integration.sh: improve variable sanity check

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 07 Jun 2021 17:06:15 +0800

ubuntu-wsl-integration (0.31.2) impish; urgency=medium

  * debian/source/format:
    - set it as a native package
  * Update Copyright and debian/copyright
  * debian/control:
    - use recent debhelper
    - fix build-Depends/Depends/Recommends
    - add long description
  * Move all scripts and configuration as part of upstream source.

 -- Patrick Wu <patrick.wu@canonical.com>  Thu, 27 May 2021 09:50:11 +0200

ubuntu-wsl-integration (0.31.1) impish; urgency=medium

  * Build System Tweaks;
  * Copyright Update.

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 25 May 2021 11:09:23 +0800

ubuntu-wsl-integration (0.31.0) impish; urgency=medium

  * Add WSLg support

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 11 May 2021 17:04:53 +0800

ubuntu-wsl-integration (0.30.0) hirsute; urgency=medium

  * TUI re-design
  * validation system inclusion

 -- Patrick Wu <patrick.wu@canonical.com>  Thu, 25 Mar 2021 21:29:09 +0800

ubuntu-wsl-integration (0.29.0) hirsute; urgency=medium

  * Expand validation of DrvFs mount option
  * Update some information
  * WSL motd news brought back

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 16 Mar 2021 15:49:53 +0800

ubuntu-wsl-integration (0.28.0) hirsute; urgency=medium

  * include validation of DrvFs mount option
  * i18n initial build system

 -- Patrick Wu <patrick.wu@canonical.com>  Thu, 04 Feb 2021 16:59:17 +0800

ubuntu-wsl-integration (0.27.2) hirsute; urgency=medium

  * debian: add source/local-options for local builds

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 01 Feb 2021 10:35:18 +0800

ubuntu-wsl-integration (0.27.1) hirsute; urgency=medium

  * update translation and WSL for submitting the package.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 29 Jan 2021 13:23:09 +0800

ubuntu-wsl-integration (0.27.0) hirsute; urgency=medium

  * Several final code refactoring.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 29 Jan 2021 10:54:34 +0800

ubuntu-wsl-integration (0.26.0) hirsute; urgency=medium

  * Improved color palette for TUI.
  * Add help for TUI.

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 25 Jan 2021 11:30:12 +0800

ubuntu-wsl-integration (0.25.0) groovy; urgency=medium

  * (Experimental) Completed initial tui implementation
  * (Experimental) working import function

 -- Patrick Wu <patrick.wu@canonical.com>  Thu, 07 Jan 2021 13:51:10 +0800

ubuntu-wsl-integration (0.24.1) groovy; urgency=medium

  * (Super Experimental) Proper footer for TUI Interface

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 22 Dec 2020 21:21:01 +0800

ubuntu-wsl-integration (0.24.0) groovy; urgency=medium

  * (Experimental) Working export function
  * (Super Experimental) Usable TUI Interface; no save feature yet
  * default system/Editor huge update

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 22 Dec 2020 16:25:52 +0800

ubuntu-wsl-integration (0.23.1) groovy; urgency=medium

  * Fix build failure due to the bad POTFILES

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 21 Dec 2020 16:46:18 +0800

ubuntu-wsl-integration (0.23.0) groovy; urgency=medium

  * (Super Experimental) Kind of working tui

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 21 Dec 2020 16:24:05 +0800

ubuntu-wsl-integration (0.22.0) groovy; urgency=medium

  * (Super Experimental) Kind of working export function

 -- Patrick Wu <patrick.wu@canonical.com>  Mon, 21 Dec 2020 15:58:55 +0800

ubuntu-wsl-integration (0.21.0) groovy; urgency=medium

  * (Super Experimental) initial TUI
  * (Super Experimental) Import and Export
  * Code refactor

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 18 Dec 2020 09:33:42 +0800

ubuntu-wsl-integration (0.20.2) groovy; urgency=medium

  * config header no longer case sensitive.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 11 Dec 2020 23:37:47 +0800

ubuntu-wsl-integration (0.20.1) groovy; urgency=medium

  * critical bug fix that cause failure in update.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 11 Dec 2020 23:37:47 +0800

ubuntu-wsl-integration (0.20.0) groovy; urgency=medium

  * update some bad tooltips.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 11 Dec 2020 12:12:25 +0800

ubuntu-wsl-integration (0.19.1) groovy; urgency=medium

  * update possible build arch.

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 24 Nov 2020 12:02:32 +0800

ubuntu-wsl-integration (0.19.0) groovy; urgency=medium

  * move motd to other packages.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 25 Sep 2020 16:23:58 +0800

ubuntu-wsl-integration (0.18.3) groovy; urgency=medium

  * Fix a build_base AttributionError.
  * update wsl-motd-news to a more current one.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 18 Sep 2020 15:11:25 +0800

ubuntu-wsl-integration (0.18.2) groovy; urgency=medium

  * Fix a i18n generation issue.

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 15 Sep 2020 23:39:09 +0800

ubuntu-wsl-integration (0.18.1) groovy; urgency=medium

  * Fix a TypeError when calling commands.
  * Small update on setup system.

 -- Patrick Wu <patrick.wu@canonical.com>  Tue, 15 Sep 2020 15:35:56 +0800

ubuntu-wsl-integration (0.18.0) groovy; urgency=medium

  * First i18n implementation (no PO/MO included);
  * inputs will be validated in "update" command,;
  * allowing wild card in "show" command;
  * always yes mode (--yes) to bypass reset all confirmation;
  * colorized STDOUT.
  * 51-wsl-motd-news: Small re-formatting.

 -- Patrick Wu <patrick.wu@canonical.com>  Fri, 21 Aug 2020 15:53:03 +0800

ubuntu-wsl-integration (0.12.0) groovy; urgency=medium

  * update logic to make the ctl works.

 -- Patrick Wu <patrick.wu@canonical.com>  Wed, 19 Aug 2020 10:56:41 +0800

ubuntu-wsl-integration (0.11.0) groovy; urgency=medium

  * several logic redesign;
  * add reset all and confirmation prompt;
  * update news content.


ubuntuwslctl/tui.py
#    ubuntuwslctl.tui - TUI
#    Copyright (C) 2021 Canonical Ltd.
#
#    Authors: Patrick Wu <patrick.wu@canonical.com>


import urwid
from ubuntuwslctl.core.decor import blank, StyledCheckBox, StyledEdit, StyledText, TuiButton, SelectableStyledText, SimpleListBox
from ubuntuwslctl.core.default import conf_def
from ubuntuwslctl.utils.helper import str2bool


class Tui:
    """
    Main class of the text-based UI for Ubuntu WSL config management
    """

    _palette = [
        ('body', '', '', 'standout'),  # body
        ('header', 'white', 'light red', 'bold', "#fff", "#e95420"),  # header
        ('footer', 'black', 'dark cyan', '', "#000", "#aea79f"),  # footer
        ('footerhlt', 'white', 'black'),  # footer highlight
        ('nv', '', 'white', 'standout', "#fff", ""),  # section nav
        ('nvfc', 'white', 'black', 'standout', "#e95420", "#000"),
        ('subttl', 'light gray', '', 'standout', "#ed764d", ''),  # section subtitle
        ('editfc', 'white', 'black', 'bold'),
        ('editbx', 'black', 'white'),
        ('editcp', '', '', 'standout'),
        ('selectable', 'white', 'black'),
        ('focus', 'black', 'light gray')
    ]

    def __init__(self, handler, color_fallback=False):
        self.handler = handler
        self.config = self.handler.get_config()
        self.navigator = []
        self.content = {}
        self.screen = urwid.raw_display.Screen()

        self.screen.set_terminal_properties(2**24)
        if color_fallback:
            self.screen.set_terminal_properties(16)
        self.screen.register_palette(self._palette)

        self.header = urwid.AttrWrap(urwid.Text(u"Ubuntu WSL Configuration UI (Experimental)", align='center'),
                                     'header')
        self.footer = urwid.AttrWrap(self._footer(), 'footer')
        self.vline = urwid.SolidFill(u'\u2502')
        self.navbox = None
        self.contentbox = None
        self.focus_content = None

        self._body_builder()
        self._loop = urwid.MainLoop(self._body, screen=self.screen,
                                    unhandled_input=self._unhandled_key)

    def _fun(self, button=None, fun=None):
        """
        Core function for different actions.
        This function can accept button callback from `button` or function name from `func`

        Args:
            button: the `urwid.Button` instance returned from a button callback.
            fun: the name of the function
        """
        if button is not None:
            fun = button.label
            fun = fun[2:].lower()
        if fun in ("", "exit"):

            raise urwid.ExitMainLoop()
        elif fun == "help":
            self._popup_constructor(fun, urwid.Text((u"Use UP/DOWN/LEFT/RIGHT arrow to navigate between "
                                                     u"option.\nUse SPACE to toggle the settings. "
                                                     u"\nUse ENTER or your mouse to press a button."), align='left'))
        elif fun == "reload":
            self._popup_constructor(fun, urwid.Text(u"Configuration Reloaded.", align='left'))
        elif fun == "save":
            for i in self.content:
                for j in self.content[i]:
                    if not hasattr(j, "get_source"):
                        continue
                    k, l, m = j.get_source()
                    n = j.get_core_value()
                    self.handler.update(k, l, m, n)
            self._popup_constructor(fun, urwid.Text(u"Saved. Restart Ubuntu to make effect.", align='left'))
        elif fun == "reset":
            def _reset(button):
                self.handler.reset_all()
                self._popup_constructor(fun, urwid.Text(u"Reset complete. Restart Ubuntu to take effect.",
                                                        align='left'))
            body = urwid.Text(u"Do you really want to reset?", align='left')
            ok_btn = urwid.AttrWrap(urwid.Button('Yes', _reset), 'selectable', 'focus')
            cc_btn = urwid.AttrWrap(urwid.Button('No', self._reload_ui), 'selectable', 'focus')
            footer = urwid.GridFlow([ok_btn, cc_btn], 10, 1, 1, 'center')
            self._popup_constructor(fun, body, footer)
        elif fun == "export":
            exp_name = urwid.Edit(u"", "")

            def _export(button):
                ef = self.handler.export_file(exp_name.edit_text)
                self._popup_constructor(fun, urwid.Text(u"Exported as {}.".format(ef),
                                                        align='left'))

            body = urwid.Pile([
                urwid.Text(u"file name to export(optional): ", align='left'),
                urwid.AttrWrap(exp_name, 'editbx', 'editfc')
            ])
            ok_btn = urwid.AttrWrap(urwid.Button('Yes', _export), 'selectable', 'focus')
            cc_btn = urwid.AttrWrap(urwid.Button('No', self._reload_ui), 'selectable', 'focus')
            footer = urwid.GridFlow([ok_btn, cc_btn], 10, 1, 1, 'center')
            self._popup_constructor(fun, body, footer)
        elif fun == "import":
            exp_name = urwid.Edit(u"", "")

            def _import(button):
                if exp_name.edit_text == "":
                    b = urwid.Text(u"No input in name, action aborted.".format(exp_name.edit_text),
                                   align='left')
                    self._popup_constructor(fun, b)
                else:
                    self.handler.import_file(exp_name.edit_text)
                    b = urwid.Text(u"{} imported. Please restart Ubuntu to take effect.".format(exp_name.edit_text),
                                align='left')
                    self._popup_constructor(fun, b)

            body = urwid.Pile([
                urwid.Text(u"file name to import: ", align='left'),
                urwid.AttrWrap(exp_name, 'editbx', 'editfc')
            ])
            ok_btn = urwid.AttrWrap(urwid.Button('Yes', _import), 'selectable', 'focus')
            cc_btn = urwid.AttrWrap(urwid.Button('No', self._reload_ui), 'selectable', 'focus')
            footer = urwid.GridFlow([ok_btn, cc_btn], 10, 1, 1, 'center')
            self._popup_constructor(fun, body, footer)
        else:  # unhandled input all went here
            self._popup_constructor(fun)

    def _footer(self):
        """
        return a footer.

        Returns:
            `urwid.GridFlow` that contains footer.
        """
        return urwid.GridFlow(
            (
                urwid.AttrWrap(TuiButton([('footerhlt', u'F1'), u'Save'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F2'), u'Reset'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F3'), u'Import'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F4'), u'Export'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F5'), u'Reload'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F6'), u'Help'], self._fun), 'footer'),
                urwid.AttrWrap(TuiButton([('footerhlt', u'F7'), u'Exit'], self._fun), 'footer')
            ),
            10, 0, 0, 'left')

    def _popup_constructor(self, fun, body=None, footer=None):
        """
        Construct a popup widget that overlays the UI.

        Args:
            fun: title of the Popup
            body: content of the Popup. Leave empty for the placeholder.
            footer: footer of the Popup. Leave empty for the single Okay button.
        """
        self._loop.widget = urwid.Overlay(self._popup_widget(fun, body, footer), self._loop.widget, align='center',
                                          valign='middle', height='pack', width=40)

    def _reload_ui(self, button):
        """
        Clear the Overlay and reload everything again.
        """
        self._body_builder()
        self._loop.widget = self._body

    def _popup_widget(self, header, body=None, footer=None):
        """
        Content of the Popup Widget.

        Args:
            header: title of the Popup
            body: content of the Popup. Leave empty for the placeholder.
            footer: footer of the Popup. Leave empty for the single Okay button.
        Returns:
            `urwid.LineBox` that hold the widget.
        """
        # Body
        if body is None:
            body = urwid.Text(('This is a placeholder text that will only be displayed '
                               'when _popup_widget.body received None. If you see this '
                               'in a production version of ubuntuwsl, Please report the '
                               'bug to WSL team at Canonical.'), align='left')

        body = urwid.Padding(body, left=1, right=1)

        # Footer
        if footer is None:
            footer = urwid.Button('Okay', self._reload_ui)
            footer = urwid.AttrWrap(footer, 'selectable', 'focus')
            footer = urwid.GridFlow([footer], 8, 1, 1, 'center')

        # Layout
        return urwid.LineBox(urwid.Pile([blank, body, blank, footer]),
                             title=header.title(), title_attr='header', title_align='center')

    def _parse_config(self):
        """
        Parse the config and place them in self.havigator and self.content accordingly.
        """
        self.navigator = []
        self.content = {}

        # Widget margin calculation
        left_margin = 4
        # for i in self.config.keys():
        #     i_tmp = self.config[i]
        #     for j in i_tmp.keys():
        #         j_tmp = i_tmp[j]
        #         for k in j_tmp.keys():
        #             if isinstance(j_tmp[k], bool) and (left_margin < 4):
        #                 left_margin = 4
        #             elif isinstance(j_tmp[k], str):
        #                 if j_tmp[k].lower() in ("yes", "no", "1", "0", "true", "false") and (left_margin < 4):
        #                     left_margin = 4
        #                 elif left_margin < len(k) + 2:
        #                     left_margin = len(k) + 2

        # Real config handling part
        for i in self.config.keys():
            if i == "time_exported":
                continue
            i_def = conf_def[i]
            self.navigator.append(urwid.AttrWrap(SelectableStyledText(i_def['_friendly_name'], assigned_value=i),
                                                 'nv', 'nvfc'))
            i_tmp = self.config[i]
            self.content[i] = []
            for j in i_tmp.keys():
                j_def = i_def[j]
                self.content[i].append(blank)
                self.content[i].append(StyledText(j_def['_friendly_name'], 'subtitle'))
                self.content[i].append(blank)
                j_tmp = i_tmp[j]
                for k in j_tmp.keys():
                    k_def = j_def[k]
                    wg = None
                    if isinstance(j_tmp[k], bool):
                        wg = StyledCheckBox(k_def['_friendly_name'], j_tmp[k],
                                                              k_def['tip'], left_margin, [i, j, k])
                    elif isinstance(j_tmp[k], str):
                        if j_tmp[k].lower() in ("yes", "no", "1", "0", "true", "false"):
                            wg = StyledCheckBox(k_def['_friendly_name'], str2bool(j_tmp[k]),
                                                                  k_def['tip'], left_margin, [i, j, k])
                        else:
                            wg = StyledEdit(k_def['_friendly_name'], j_tmp[k],
                                                              k_def['tip'], left_margin, [i, j, k])
                    self.content[i].append(wg)
                    self.content[i].append(blank)

    def _content_validation(self):
        """
        Validate the content in current widget as the focus moved away.
        """
        if self.focus_content is None:
            self.focus_content = self.contentbox.focus_position
            return
        if self.focus_content > len(self.contentbox.walker):
            return
        realc = self.contentbox.walker[self.focus_content]
        #self._popup_constructor(u"Debug", urwid.Text(str(realc)))
        self.focus_content = self.contentbox.focus_position
        if not hasattr(realc, "get_source"):
            return
        k, l, m = realc.get_source()
        n = realc.get_core_value()
        a, b = self.handler.validate(k, l, m, n)
        if not a:
            def _better_reload(button):
                self._nav_update(forced_value=k)
                self.contentbox.set_focus(self.focus_content)
            footer = urwid.AttrWrap(urwid.Button('Okay', _better_reload), 'selectable', 'focus')
            footer = urwid.GridFlow([footer], 8, 1, 1, 'center')
            self._popup_constructor(u"Validation Error", urwid.Text(b), footer)

    def _nav_update(self, forced_value=None):
        """
        Update the navigation location

        Args:
            forced_value: force to set the navigation to a section. default is None.
        """
        def_nav = self.navbox.get_focus()[0].get_assigned_value()
        if forced_value is not None:
            def_nav = forced_value
        self.contentbox = SimpleListBox(self.content[def_nav])
        content = urwid.Columns([self.navbox,('fixed', 1, self.vline), ('weight', 2, self.contentbox)])
        urwid.connect_signal(self.contentbox.walker, 'modified', self._content_validation)
        self._loop.widget = urwid.Frame(urwid.AttrWrap(content, 'body'), header=self.header, footer=self.footer)

    def _body_builder(self):
        """
        Allows building the body.
        Also used in page refreshing after a reset and when pressing reload botton
        """
        self._parse_config()

        self.navbox = SimpleListBox(self.navigator)
        def_nav = self.navbox.get_focus()[0].get_assigned_value()
        self.contentbox = SimpleListBox(self.content[def_nav])
        content = urwid.Columns([self.navbox,('fixed', 1, self.vline), ('weight', 2, self.contentbox)])
        urwid.connect_signal(self.navbox.walker, 'modified', self._nav_update)
        urwid.connect_signal(self.contentbox.walker, 'modified', self._content_validation)
        self._body = urwid.Frame(urwid.AttrWrap(content, 'body'), header=self.header, footer=self.footer)

    def _unhandled_key(self, key):
        """
        handle keys
        """
        if key in ('f7', 'ctrl c', 'esc'):
            self._fun(fun='exit')
        elif key == 'f6':
            self._fun(fun='help')
        elif key == 'f5':
            self._fun(fun='reload')
        elif key == 'f4':
            self._fun(fun='export')
        elif key == 'f3':
            self._fun(fun='import')
        elif key == 'f2':
            self._fun(fun='reset')
        elif key == 'f1':
            self._fun(fun='save')

    def run(self):
        """
        Start the UI
        """
        self._loop.run()
